### PR TITLE
Add support for system-level routing

### PR DESCRIPTION
**Pull request by abarth**
Now you can supply a route argument to flutter start:

  `$ flutter start --route /foo`

which will start the app at /foo instead of /.

### CODE REVIEW

**Comment by abarth** [comment]
@Hixie @chinmaygarde

File: packages/flutter/lib/src/material/material_app.dart [hunk 1]
```diff
@@ -68,30 +68,32 @@ class _MaterialAppState extends State<MaterialApp> {
...
 
   void dispose() {
-    WidgetFlutterBinding.instance.removeEventListener(_backHandler);
-    FlutterBinding.instance.removeMetricListener(_metricHandler);
+    FlutterBinding.instance.removeObserver(this);
     super.dispose();
   }
 
-  void _backHandler(InputEvent event) {
+  void onPushRoute(String route) {
```
Comment (Hixie) [inline packages/flutter/lib/src/material/material_app.dart]: We've been using onFoo() for the fields you set, and handleFoo() for the handlers.

File: packages/flutter/lib/src/material/material_app.dart [hunk 2]
```diff
@@ -68,30 +68,32 @@ class _MaterialAppState extends State<MaterialApp> {
...
+    print("onPushRoute $route");
```
Comment (Hixie) [inline packages/flutter/lib/src/material/material_app.dart]: print

File: packages/flutter/lib/src/material/material_app.dart [hunk 3]
```diff
@@ -68,30 +68,32 @@ class _MaterialAppState extends State<MaterialApp> {
...
+    NavigatorState navigator = _navigator.currentState;
+    navigator.pushNamed(route);
```
Comment (Hixie) [inline packages/flutter/lib/src/material/material_app.dart]: you don't assert here, but do below. should be consistent.

File: packages/flutter/lib/src/material/material_app.dart [hunk 4]
```diff
@@ -68,30 +68,32 @@ class _MaterialAppState extends State<MaterialApp> {
...
-      NavigatorState navigator = _navigator.currentState;
-      assert(navigator != null);
-      if (navigator.hasPreviousRoute)
-        navigator.pop();
-      else
-        activity.finishCurrentActivity();
-    }
+    NavigatorState navigator = _navigator.currentState;
+    assert(navigator != null);
+    if (navigator.hasPreviousRoute)
```
Comment (Hixie) [inline packages/flutter/lib/src/material/material_app.dart]: this will merge conflict

File: packages/flutter/lib/src/rendering/binding.dart [hunk 5]
```diff
@@ -192,23 +196,15 @@ class FlutterBinding extends HitTestTarget {
...
-    assert(eventType == 'back');
-    InputEvent ourEvent = new InputEvent(
-      type: eventType,
-      timeStamp: new Duration(microseconds: timeStamp.round())
-    );
-    for (EventListener listener in _eventListeners)
-      listener(ourEvent);
+  void _handlePopRoute() {
+    for (BindingObserver observer in _observers)
+      observer.onPopRoute();
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/binding.dart]: That's scary. What if they all do something?

Maybe there should only be one observer? (Like for Navigator currently?)

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 6]
```diff
@@ -72,7 +74,9 @@ class NavigatorState extends State<Navigator> {
     super.initState();
     assert(config.observer == null || config.observer.navigator == null);
     config.observer?._navigator = this;
-    push(config.onGenerateRoute(new NamedRouteSettings(name: Navigator.defaultRouteName)));
+    push(config.onGenerateRoute(new NamedRouteSettings(
+      name: config.initialRoute ?? Navigator.defaultRouteName
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/navigator.dart]: push the default up to the constructor argument

**Comment by Hixie** [comment]
LGTM with minor comments

**Pull request merged**